Flaky test: ZooKeeperRegistrationTest.testConnectionRecovery

```
ZooKeeperRegistrationTest.testConnectionRecovery: FAILURE
com.linecorp.armeria.server.zookeeper.ZooKeeperRegistrationTest > testConnectionRecovery FAILED
    java.lang.AssertionError: expected:<Expired> but was:<Disconnected>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:834)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at com.linecorp.armeria.server.zookeeper.ZooKeeperRegistrationTest.testConnectionRecovery(ZooKeeperRegistrationTest.java:137)
```

--- a/build.gradle
+++ b/build.gradle
@@ -764,6 +764,11 @@ configure(javaProjects) {
                         case 'com.linecorp.armeria.client.zookeeper.EndpointGroupTest':
                             lines.removeIf { it.contains('ServerCnxn$EndOfStreamException') }
                             break
+                        case 'com.linecorp.armeria.server.zookeeper.ZooKeeperRegistrationTest':
+                            lines.removeIf { it.contains('ServerCnxn$EndOfStreamException') }
+                            lines.removeIf { it.contains('KeeperException$ConnectionLossException') }
+                            lines.removeIf { it.contains('ZooKeeperException: Failed to notify ZooKeeper listener') }
+                            break
                     }
 
                     printBuffer = !lines.isEmpty()

--- a/core/src/test/java/com/linecorp/armeria/server/ServerTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/ServerTest.java
@@ -201,7 +201,7 @@ public class ServerTest {
             }
             long elapsedTimeMillis = TimeUnit.MILLISECONDS.convert(
                     System.nanoTime() - connectedNanos, TimeUnit.NANOSECONDS);
-            assertThat(elapsedTimeMillis).isGreaterThanOrEqualTo(idleTimeoutMillis);
+            assertThat(elapsedTimeMillis).isGreaterThan((long) (idleTimeoutMillis * 0.9));
         }
     }
 

--- a/dependencies.yml
+++ b/dependencies.yml
@@ -123,7 +123,12 @@ org.apache.tomcat.embed:
   tomcat-embed-el: { version: *TOMCAT_VERSION }
 
 org.apache.zookeeper:
-  zookeeper: { version: '3.4.10' }
+  zookeeper:
+    version: '3.4.10'
+    exclusions:
+    - jline:jline
+    - log4j:log4j
+    - org.slf4j:slf4j-log4j12
 
 org.assertj:
   assertj-core: { version: '3.8.0' }

--- a/zookeeper/src/test/java/com/linecorp/armeria/client/zookeeper/EndpointGroupTest.java
+++ b/zookeeper/src/test/java/com/linecorp/armeria/client/zookeeper/EndpointGroupTest.java
@@ -18,12 +18,13 @@ package com.linecorp.armeria.client.zookeeper;
 import static com.linecorp.armeria.client.zookeeper.ZooKeeperEndpointGroup.Mode.IN_CHILD_NODES;
 import static com.linecorp.armeria.client.zookeeper.ZooKeeperEndpointGroup.Mode.IN_NODE_VALUE;
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.Assert.assertEquals;
+import static org.awaitility.Awaitility.await;
 import static org.junit.Assert.fail;
 
 import java.util.Collection;
 import java.util.Collections;
 import java.util.EnumSet;
+import java.util.List;
 import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 
@@ -54,13 +55,11 @@ import zookeeperjunit.ZooKeeperAssert;
 
 @RunWith(Parameterized.class)
 public class EndpointGroupTest extends TestBase implements ZooKeeperAssert, OptionAssert {
-    protected static final KeeperState[] expectedStates = {
-            KeeperState.Disconnected, KeeperState.Expired,
-            KeeperState.SyncConnected, KeeperState.Disconnected
-    };
+
     @Parameter
     @SuppressWarnings("VisibilityModifier")
     public Mode mode;
+
     private ZooKeeperEndpointGroup endpointGroup;
 
     @Parameters
@@ -111,8 +110,6 @@ public class EndpointGroupTest extends TestBase implements ZooKeeperAssert, Opti
     public void testUpdateEndpointGroup() throws Throwable {
         Set<Endpoint> expected = ImmutableSet.of(Endpoint.of("127.0.0.1", 8001, 2),
                                                  Endpoint.of("127.0.0.1", 8002, 3));
-        CountDownLatch latch = new CountDownLatch(1);
-        endpointGroup.addListener(l -> latch.countDown());
         switch (mode) {
             case IN_NODE_VALUE:
                 setNodeValue(NodeValueCodec.DEFAULT.encodeAll(expected));
@@ -130,8 +127,9 @@ public class EndpointGroupTest extends TestBase implements ZooKeeperAssert, Opti
             zk.sync(zNode, (rc, path, ctx) -> {
             }, null);
         }
-        latch.await();
-        assertThat(endpointGroup.endpoints()).hasSameElementsAs(expected);
+
+        final Set<Endpoint> finalExpected = expected;
+        await().untilAsserted(() -> assertThat(endpointGroup.endpoints()).hasSameElementsAs(finalExpected));
     }
 
     @Test
@@ -148,9 +146,34 @@ public class EndpointGroupTest extends TestBase implements ZooKeeperAssert, Opti
         latch.await();
         //once connected, close the new handler to cause the original handler session expire
         zkHandler2.close();
-        for (KeeperState state : expectedStates) {
-            assertEquals(state, endpointGroup.stateQueue().take());
+
+        // Ensure the state transition went as expected.
+        final List<KeeperState> actualStates = takeAllStates(endpointGroup.stateQueue());
+        int i = 0;
+
+        // Expect the initial disconnection events.
+        int numDisconnected = 0;
+        for (; i < actualStates.size(); i++) {
+            if (actualStates.get(i) != KeeperState.Disconnected) {
+                break;
+            }
+            numDisconnected++;
         }
+        assertThat(numDisconnected).isGreaterThan(0);
+
+        assertThat(actualStates.get(i++)).isEqualTo(KeeperState.Expired);
+        assertThat(actualStates.get(i++)).isEqualTo(KeeperState.SyncConnected);
+
+        // Expect the last disconnection events.
+        numDisconnected = 0;
+        for (; i < actualStates.size(); i++) {
+            if (actualStates.get(i) != KeeperState.Disconnected) {
+                break;
+            }
+            numDisconnected++;
+        }
+        assertThat(numDisconnected).isGreaterThan(0);
+
         testGetEndpointGroup();
     }
 

--- a/zookeeper/src/test/java/com/linecorp/armeria/client/zookeeper/TestBase.java
+++ b/zookeeper/src/test/java/com/linecorp/armeria/client/zookeeper/TestBase.java
@@ -15,9 +15,19 @@
  */
 package com.linecorp.armeria.client.zookeeper;
 
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
 import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
 import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
 
+import org.apache.zookeeper.Watcher.Event.KeeperState;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
 import org.junit.ClassRule;
@@ -36,31 +46,61 @@ public class TestBase {
     protected static final Logger logger = LoggerFactory.getLogger(EndpointGroupTest.class);
     protected static final String zNode = "/testEndPoints";
     protected static final int sessionTimeout = 20000;
-    protected static final Set<Endpoint> sampleEndpoints = ImmutableSet.of(Endpoint.of("127.0.0.1", 1234, 2),
-                                                                           Endpoint.of("127.0.0.1", 2345, 4),
-                                                                           Endpoint.of("127.0.0.1", 3456, 2));
+    protected static final Set<Endpoint> sampleEndpoints;
+
+    static {
+        final int[] ports = unusedPorts(3);
+        sampleEndpoints = ImmutableSet.of(Endpoint.of("127.0.0.1", ports[0], 2),
+                                          Endpoint.of("127.0.0.1", ports[1], 4),
+                                          Endpoint.of("127.0.0.1", ports[2], 2));
+    }
+
     private static final Duration duration = Duration.ofSeconds(10);
     @ClassRule
     public static final TemporaryFolder ROOT_FOLDER = new TemporaryFolder();
     private static ZKInstance zkInstance;
 
     @BeforeClass
-    public static void start() {
-        try {
-            zkInstance = ZKFactory.apply().withRootDir(ROOT_FOLDER.newFolder("zookeeper")).create();
-            zkInstance.start().result(duration);
-        } catch (Throwable throwable) {
-            throw new IllegalStateException(throwable);
-        }
+    public static void start() throws Throwable {
+        zkInstance = ZKFactory.apply().withRootDir(ROOT_FOLDER.newFolder("zookeeper")).create();
+        zkInstance.start().result(duration);
     }
 
     @AfterClass
-    public static void stop() {
-        try {
-            zkInstance.stop().ready(duration);
-        } catch (Exception e) {
-            logger.warn("Failed to stop the ZooKeeper zkInstance", e);
+    public static void stop() throws Throwable {
+        zkInstance.stop().ready(duration);
+    }
+
+    protected static List<KeeperState> takeAllStates(BlockingQueue<KeeperState> queue) throws Exception {
+        final List<KeeperState> actualStates = new ArrayList<>();
+        for (;;) {
+            final KeeperState state = queue.poll(3, TimeUnit.SECONDS);
+            if (state == null) {
+                break;
+            }
+            actualStates.add(state);
         }
+        return actualStates;
+    }
+
+    private static int[] unusedPorts(int numPorts) {
+        final int[] ports = new int[numPorts];
+        final Random random = ThreadLocalRandom.current();
+        for (int i = 0; i < numPorts; i++) {
+            for (;;) {
+                final int candidatePort = random.nextInt(64512) + 1024;
+                try (ServerSocket ss = new ServerSocket()) {
+                    ss.bind(new InetSocketAddress("127.0.0.1", candidatePort));
+                    ports[i] = candidatePort;
+                    break;
+                } catch (IOException e) {
+                    // Port in use or unable to bind.
+                    continue;
+                }
+            }
+        }
+
+        return ports;
     }
 
     public ZKInstance instance() {

--- a/zookeeper/src/test/java/com/linecorp/armeria/server/zookeeper/ZooKeeperRegistrationTest.java
+++ b/zookeeper/src/test/java/com/linecorp/armeria/server/zookeeper/ZooKeeperRegistrationTest.java
@@ -15,8 +15,8 @@
  */
 package com.linecorp.armeria.server.zookeeper;
 
-import static org.assertj.core.api.Assertions.fail;
-import static org.junit.Assert.assertEquals;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.fail;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -24,10 +24,8 @@ import java.util.concurrent.CountDownLatch;
 
 import org.apache.zookeeper.Watcher.Event.KeeperState;
 import org.apache.zookeeper.ZooKeeper;
-import org.assertj.core.api.Assertions;
 import org.junit.After;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Test;
 
 import com.linecorp.armeria.client.Endpoint;
@@ -50,40 +48,40 @@ import zookeeperjunit.CloseableZooKeeper;
 import zookeeperjunit.ZooKeeperAssert;
 
 public class ZooKeeperRegistrationTest extends TestBase implements ZooKeeperAssert, OptionAssert {
-    protected static final KeeperState[] expectedStates = {
-            KeeperState.Disconnected, KeeperState.Expired,
-            KeeperState.SyncConnected, KeeperState.SyncConnected, KeeperState.Disconnected
-    };
-    List<Server> servers;
-    List<ZooKeeperRegistration> zkConnectors;
-    List<ZooKeeperUpdatingListener> listeners;
+
+    private List<Server> servers;
+    private List<ZooKeeperRegistration> zkConnectors;
+    private List<ZooKeeperUpdatingListener> listeners;
 
     @Before
-    public void startServer() {
+    public void startServers() throws Exception {
         servers = new ArrayList<>();
         zkConnectors = new ArrayList<>();
         listeners = new ArrayList<>();
-        try {
-            for (Endpoint endpoint : sampleEndpoints) {
-                ServerBuilder sb = new ServerBuilder();
-                Server server = sb.service("/", new EchoService()).port(endpoint.port(),
-                                                                        SessionProtocol.HTTP).build();
-                ZooKeeperUpdatingListener listener;
-                listener = new ZooKeeperUpdatingListener(instance().connectString().get(), zNode,
-                                                         sessionTimeout,
-                                                         endpoint);
-                server.addListener(listener);
-                server.start().get();
-                listeners.add(listener);
-                zkConnectors.add(listener.getConnector());
-                servers.add(server);
-            }
-        } catch (Exception e) {
-            fail(e.getMessage());
+
+        for (Endpoint endpoint : sampleEndpoints) {
+            Server server = new ServerBuilder().port(endpoint.port(), SessionProtocol.HTTP)
+                                               .service("/", new EchoService())
+                                               .build();
+            ZooKeeperUpdatingListener listener;
+            listener = new ZooKeeperUpdatingListener(instance().connectString().get(), zNode,
+                                                     sessionTimeout, endpoint);
+            server.addListener(listener);
+            server.start().join();
+            listeners.add(listener);
+            zkConnectors.add(listener.getConnector());
+            servers.add(server);
         }
     }
 
-    @Test
+    @After
+    public void stopServers() throws Exception {
+        if (servers != null) {
+            servers.forEach(s -> s.stop().join());
+        }
+    }
+
+    @Test(timeout = 10000)
     public void testServerNodeCreateAndDelete() {
         //all servers start and with zNode created
         sampleEndpoints.forEach(
@@ -92,10 +90,9 @@ public class ZooKeeperRegistrationTest extends TestBase implements ZooKeeperAsse
             try {
                 sampleEndpoints.forEach(endpoint -> {
                     try {
-                        Assertions.assertThat(NodeValueCodec.DEFAULT.decode(
-                                zkClient.getData(zNode + '/' + endpoint.host() + '_' + endpoint.port()).get()))
-                                  .isEqualTo(
-                                          endpoint);
+                        assertThat(NodeValueCodec.DEFAULT.decode(zkClient.getData(
+                                zNode + '/' + endpoint.host() + '_' + endpoint.port()).get()))
+                                .isEqualTo(endpoint);
                     } catch (Throwable throwable) {
                         fail(throwable.getMessage());
                     }
@@ -118,14 +115,14 @@ public class ZooKeeperRegistrationTest extends TestBase implements ZooKeeperAsse
         }
     }
 
-    @Ignore  // FIXME: https://github.com/line/armeria/issues/477
-    @Test
+    @Test(timeout = 10000)
     public void testConnectionRecovery() throws Exception {
         ZooKeeperRegistration zkConnector = zkConnectors.get(0);
         zkConnector.enableStateRecording();
         ZooKeeper zkHandler1 = zkConnector.underlyingClient();
         CountDownLatch latch = new CountDownLatch(1);
         ZooKeeper zkHandler2;
+
         //create a new handler with the same sessionId and password
         zkHandler2 = new ZooKeeper(instance().connectString().get(), sessionTimeout, event -> {
             if (event.getState() == KeeperState.SyncConnected) {
@@ -133,24 +130,48 @@ public class ZooKeeperRegistrationTest extends TestBase implements ZooKeeperAsse
             }
         }, zkHandler1.getSessionId(), zkHandler1.getSessionPasswd());
         latch.await();
+
         //once connected, close the new handler to cause the original handler session expire
         zkHandler2.close();
-        for (KeeperState state : expectedStates) {
-            assertEquals(state, zkConnector.stateQueue().take());
+
+        // Ensure the state transition went as expected.
+        final List<KeeperState> actualStates = takeAllStates(zkConnector.stateQueue());
+        int i = 0;
+
+        // Expect the initial disconnection events.
+        int numDisconnected = 0;
+        for (; i < actualStates.size(); i++) {
+            if (actualStates.get(i) != KeeperState.Disconnected) {
+                break;
+            }
+            numDisconnected++;
         }
+        assertThat(numDisconnected).isGreaterThan(0);
+
+        assertThat(actualStates.get(i++)).isEqualTo(KeeperState.Expired);
+        assertThat(actualStates.get(i++)).isEqualTo(KeeperState.SyncConnected);
+
+        // Expect the optional disconnection events.
+        for (; i < actualStates.size(); i++) {
+            if (actualStates.get(i) != KeeperState.Disconnected) {
+                break;
+            }
+        }
+
+        assertThat(actualStates.get(i++)).isEqualTo(KeeperState.SyncConnected);
+
+        // Expect the last disconnection events.
+        numDisconnected = 0;
+        for (; i < actualStates.size(); i++) {
+            if (actualStates.get(i) != KeeperState.Disconnected) {
+                break;
+            }
+            numDisconnected++;
+        }
+        assertThat(numDisconnected).isGreaterThan(0);
+
         //connection will recover and our ZooKeeper node also exists
         testServerNodeCreateAndDelete();
-    }
-
-    @After
-    public void stopServer() {
-        try {
-            for (Server server : servers) {
-                server.stop().get();
-            }
-        } catch (Exception ex) {
-            fail(ex.getMessage());
-        }
     }
 
     private static class EchoService extends AbstractHttpService {

